Ghidra/Features/Decompiler/src/main/java/ghidra/app/decompiler/DecompileCallback.java

```diff
@@ -1316,15 +1316,14 @@ public StringData getStringData(String addrString, String dtName, String dtId) {
 		}
 		Data data = program.getListing().getDataContaining(addr);
 		Settings settings = SettingsImpl.NO_SETTINGS;
-		AbstractStringDataType dataType = null;
 		StringDataInstance stringInstance = null;
 		int length = 0;
 		if (data != null) {
-			if (data.getDataType() instanceof AbstractStringDataType) {
-				// There is already a string here.  Use its configuration to
-				// set up the StringDataInstance
+			if (data.getDataType() instanceof AbstractStringDataType
```

Comment:
```suggestion
			if (data.hasStringValue())
```
I believe this method is more appropriate here.

ps I'm on my phone so line selections may be wierd.

Reply:
Maybe it is more appropriate. But for now I can not be 100% sure if `data.hasStringValue()` returns true, the only possible data types are `AbstractStringDataType` and `Array`.

And later, the `getStringDataInstance` method are not abstracted out of `DataType`, i.e, there's no interface contains such a method.
Thus, current implementation checks the type with `instanceof` one by one (this it the implementation before this patch anyway).
If we use `hasStringValue()` here, I suppose we have two options:

1. figure out what are the potential types and type check and cast them one by one then call `getStringDataInstance` method.
2. do some more abstraction, use interface like `StringableDataType` and contains method of `getStringDataInstance`.

And for now, I'm not confident enough I can do either without any help. (I'm also working on some more fixings related to better Rust decompilation. I tend to make them work first)

Reply:
Try the following. It appears that the work is already done in StringDataInstance.
You can actually skip the entire `if (data == null)` block and instead do the following:
```java
StringDataInstance stringInstance = StringDataInstance.getStringDataInstance(data);
if (stringInstance == StringDataInstance.NULL_INSTANCE) {
    // previous if (data != null) else block contents L1348-L1377
}
```

As a side note, when you have a `Data` instance and `Data.hasStringValue()` returns `true` you can do the following:
```java
public static String getStringFromData(Data data) {
    if (data.hasStringValue()) {
        return (String) data.getValue();
    }
    return "";
}
```

Reply:
Thanks for the advice! I have pushed the new code adopting that option.

Reply:
I'm terribly sorry. I made a mistake. Removing the `if (data != null)` completely breaks what the patch was intended to fix. IT should go back to how that was. However, it can be skipped if `data.getAddress().equals(addr)`.

Reply:
Ah! I just noticed it! I will make it back. You are right, current patch introduces incorrect starting address problem